src/IotTelemetrySimulator/TelemetryValues.cs

```diff
@@ -123,26 +130,21 @@ public TelemetryValues(IList<TelemetryVariable> variables)
         /// Removes non-used variables in a sequence.
         /// This way we can keep the a counter variable incrementally correctly if the sequence did not use it in current iteration.
         /// </summary>
-        private static void ResetNonUsedReferencedVariables(
+        private static void ResetNotUsedReferencedVariables(
             Dictionary<string, object> previous,
             Dictionary<string, object> next,
-            TelemetryVariable sequenceVariable,
-            string usedVariable)
+            IEnumerable<string> notUserVariables)
         {
-            var referencedVariables = sequenceVariable.GetReferenceVariableNames();
-            foreach (var referencedVariable in referencedVariables)
+            foreach (var notUsedVariable in notUserVariables)
             {
-                if (referencedVariable != usedVariable)
+                // Restore it from the previous value.
+                if (previous != null && previous.TryGetValue(notUsedVariable, out var previousValue))
```

Comment:
That would require us to have two loops, no?

Reply:
No, I mean:

```cs
if (previous is null) {
    return;
}
```

Reply:
But that wouldn't cover the scenario of a first interaction where a counter is not used. In this case we should remove it from the "next" so that it can start from the beginning the next time. Right?

Reply:
Ah, yes, I forgot there's `else`. Nevermind :)